### PR TITLE
Implement inactivity timeout in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
This is a port of the C++ implementation to C#.

### CODE REVIEW

File: csharp/test/Ice/inactivityTimeout/AllTests.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    private static async Task testServerInactivityTimeout(string proxyString, Properties properties, TextWriter output)
+    {
+        output.Write("testing that the server side inactivity timeout shuts down the connection... ");
+        output.Flush();
+
+        // Create a new communicator with the desired properties.
+        properties = properties.Clone();
+        properties.setProperty("Ice.Connection.InactivityTimeout", "3");
```
Comment (pepone) [inline csharp/test/Ice/inactivityTimeout/AllTests.cs:48]: Shouldn't we set it to `5` as mentioned in the comment below?

```suggestion
        properties.setProperty("Ice.Connection.InactivityTimeout", "5");
```

File: csharp/src/Ice/ConnectionI.cs [hunk 2]
```diff
@@ -2085,8 +2094,44 @@ private int sendNextMessage(out Queue<OutgoingMessage> callbacks)
 
     private int sendMessage(OutgoingMessage message)
     {
+        Debug.Assert(_state >= StateActive);
         Debug.Assert(_state < StateClosed);
 
+        bool isHeartbeat = message.stream.getBuffer().b.get(8) == Protocol.validateConnectionMsg;
+        if (!isHeartbeat)
+        {
+            cancelInactivityTimer();
```
Comment (pepone) [inline csharp/src/Ice/ConnectionI.cs:2103]: Can we unschedule without disposing of the timer in this case?

File: csharp/src/Ice/ConnectionI.cs [hunk 3]
```diff
@@ -2816,6 +2917,10 @@ internal void completed(LocalException ex)
 
     private int _upcallCount;
```
Comment (pepone) [inline csharp/src/Ice/ConnectionI.cs:2918]: Not related to the PR: Can you add a comment to clarify `_upcallCount`?

**Review by pepone — APPROVED** [review]

File: csharp/test/Ice/inactivityTimeout/AllTests.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/test/Ice/inactivityTimeout/AllTests.cs:48]: Fixed!

File: csharp/src/Ice/ConnectionI.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionI.cs:2103]: That's not the logic on the code here. Unschedule = dispose and set to null. Schedule = create a new timer.
And we rely on this logic in inactivityCheck when we compare inactivityTimer with _inacvityTimer.

The scheduling is very infrequent so there is no need to worry about reusing the timer object.

File: csharp/src/Ice/ConnectionI.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionI.cs:2918]: Yes, done.

**Pull request merged**